File: tests/test-list.c
#include <stdio.h>

struct list_head {
	struct list_head *next;
	struct list_head *prev;
};

extern struct list_head normal_head;

struct list_head short_list[] = {
	{
		.next = &short_list[1],
		.prev = &normal_head,
	},
	{
		.next = &short_list[2],
		.prev = &short_list[0],
	},
	{
		.next = &short_list[3],
		.prev = &short_list[1],
	},
	{
		.next = &short_list[4],
		.prev = &short_list[2],
	},
	{
		.next = &normal_head,
		.prev = &short_list[3],
	},
};

struct list_head normal_head = {
	.next = &short_list[0],
	.prev = &short_list[4],
};

extern struct list_head cycle_head;
struct list_head short_list_with_cycle[] = {
	{
		.next = &short_list_with_cycle[1],
		.prev = &cycle_head,
	},
	{
		.next = &short_list_with_cycle[2],
		.prev = &short_list_with_cycle[0],
	},
	{
		.next = &short_list_with_cycle[3],
		.prev = &short_list_with_cycle[1],
	},
	{
		.next = &short_list_with_cycle[1],
		.prev = &short_list_with_cycle[2],
	},
	{
		.next = &cycle_head,
		.prev = &short_list_with_cycle[3],
	},
};

struct list_head cycle_head = {
	.next = &short_list_with_cycle[0],
	.prev = &short_list_with_cycle[4],
};

extern struct list_head bad_list_head;
struct list_head short_list_with_bad_prev[] = {
	{
		.next = &short_list_with_bad_prev[1],
		.prev = &bad_list_head,
	},
	{
		.next = &short_list_with_bad_prev[2],
		.prev = &short_list_with_bad_prev[0],
	},
	{
		.next = &short_list_with_bad_prev[3],
		.prev = &short_list_with_bad_prev[0],
	},
	{
		.next = &short_list_with_bad_prev[1],
		.prev = &short_list_with_bad_prev[2],
	},
	{
		.next = &bad_list_head,
		.prev = &short_list_with_bad_prev[3],
	},
};

struct list_head bad_list_head = {
	.next = &short_list[0],
	.prev = &short_list[4],
};

struct container {
	unsigned long someval;
	struct list_head list;
};

extern struct list_head good_container_list;
struct container good_containers[] = {
	{
		.someval = 0xdead0000,
		.list = {
			.next = &good_containers[1].list,
			.prev = &good_container_list,
		},
	},
	{
		.someval = 0xdead0001,
		.list = {
			.next = &good_containers[2].list,
			.prev = &good_containers[0].list,
		},
	},
	{
		.someval = 0xdead0002,
		.list = {
			.next = &good_containers[3].list,
			.prev = &good_containers[1].list,
		},
	},
	{
		.someval = 0xdead0003,
		.list = {
			.next = &good_containers[4].list,
			.prev = &good_containers[2].list,
		},
	},
	{
		.someval = 0xdead0004,
		.list = {
			.next = &good_container_list,
			.prev = &good_containers[3].list,
		},
	},
};

struct list_head good_container_list = {
	.next = &good_containers[0].list,
	.prev = &good_containers[4].list,
};

extern struct list_head cycle_container_list;
struct container cycle_containers[] = {
	{
		.someval = 0xdead0000,
		.list = {
			.next = &cycle_containers[1].list,
			.prev = &cycle_container_list,
		},
	},
	{
		.someval = 0xdead0001,
		.list = {
			.next = &cycle_containers[2].list,
			.prev = &cycle_containers[0].list,
		},
	},
	{
		.someval = 0xdead0002,
		.list = {
			.next = &cycle_containers[3].list,
			.prev = &cycle_containers[1].list,
		},
	},
	{
		.someval = 0xdead0003,
		.list = {
			.next = &cycle_containers[1].list,
			.prev = &cycle_containers[2].list,
		},
	},
	{
		.someval = 0xdead0004,
		.list = {
			.next = &cycle_container_list,
			.prev = &cycle_containers[3].list,
		},
	},
};

struct list_head cycle_container_list = {
	.next = &cycle_containers[0].list,
	.prev = &cycle_containers[4].list,
};

extern struct list_head bad_container_list;
struct container bad_containers[] = {
	{
		.someval = 0xdead0000,
		.list = {
			.next = &bad_containers[1].list,
			.prev = &bad_container_list,
		},
	},
	{
		.someval = 0xdead0001,
		.list = {
			.next = &bad_containers[2].list,
			.prev = &bad_containers[0].list,
		},
	},
	{
		.someval = 0xdead0002,
		.list = {
			.next = &bad_containers[3].list,
			.prev = &bad_containers[1].list,
		},
	},
	{
		.someval = 0xdead0003,
		.list = {
			.next = &bad_containers[4].list,
			.prev = &bad_containers[1].list,
		},
	},
	{
		.someval = 0xdead0004,
		.list = {
			.next = &bad_container_list,
			.prev = &bad_containers[3].list,
		},
	},
};

struct list_head bad_container_list = {
	.next = &bad_containers[0].list,
	.prev = &bad_containers[4].list,
};

struct list_head bad_next_ptr_list = {
	.next = (struct list_head *)0xdeadbeef,
	.prev = &bad_next_ptr_list,
};

struct list_head bad_prev_ptr_list = {
	.next = &bad_prev_ptr_list,
	.prev = (struct list_head *)0xdeadbeef,
};

int
main(void)
{
	struct list_head lh = { NULL, NULL };
	printf("normal_head = %p\n", &normal_head);
	printf("short_list = %p\n", &short_list);
	return 0;
}

File: tests/test_list.py
# -*- coding: utf-8 -*-
# vim:set shiftwidth=4 softtabstop=4 expandtab textwidth=79:

import unittest
import gdb

from crash.types.list import list_for_each, list_for_each_entry
from crash.types.list import ListCycleError, CorruptListError

def get_symbol(name):
    return gdb.lookup_symbol(name, None)[0].value()

class TestList(unittest.TestCase):
    def setUp(self):
        gdb.execute("file tests/test-list")
        self.list_head = gdb.lookup_type("struct list_head")

    def tearDown(self):
        gdb.execute("file")

    def test_none_list(self):
        count = 0
        with self.assertRaises(TypeError):
            for node in list_for_each(None):
                count += 1

    def test_invalid_value(self):
        count = 0
        gdbtype = gdb.lookup_type('unsigned int')
        with self.assertRaises(TypeError):
            for node in list_for_each(gdbtype):
                count += 1

    def test_invalid_value_pointer(self):
        count = 0
        gdbtype = gdb.lookup_type('unsigned int').pointer()
        with self.assertRaises(TypeError):
            for node in list_for_each(gdbtype):
                count += 1

    def test_bad_next_pointer_list(self):
        head = get_symbol("bad_next_ptr_list")
        count = 0

        with self.assertRaises(BufferError):
            for node in list_for_each(head):
                count += 1

#    Not a failure yet - we don't iterate the list backward
#    def test_bad_prev_pointer_list(self):
#        head = get_symbol("bad_prev_ptr_list")
#        count = 0
#
#        with self.assertRaises(BufferError):
#            for node in list_for_each(head):
#                count += 1

    def test_normal_list(self):
        normal_list = get_symbol("normal_head")
        short_list = get_symbol("short_list")
        expected_count = short_list.type.sizeof // short_list[0].type.sizeof
        count = 0
        for node in list_for_each(normal_list):
            count += 1

        self.assertTrue(count == expected_count)

    def test_cycle_list(self):
        normal_list = get_symbol("cycle_head")
        short_list = get_symbol("short_list_with_cycle")
        expected_count = short_list.type.sizeof // short_list[0].type.sizeof
        count = 0
        with self.assertRaises(ListCycleError):
            for node in list_for_each(normal_list, exact_cycles=True):
                count += 1

    def test_corrupt_list(self):
        normal_list = get_symbol("bad_list_head")
        short_list = get_symbol("short_list_with_bad_prev")
        expected_count = short_list.type.sizeof // short_list[0].type.sizeof
        count = 0
        with self.assertRaises(CorruptListError):
            for node in list_for_each(normal_list, exact_cycles=True,
                                      print_broken_links=False):
                count += 1

    def test_normal_container_list_with_string(self):
        normal_list = get_symbol("good_container_list")
        short_list = get_symbol("good_containers")
        expected_count = short_list.type.sizeof // short_list[0].type.sizeof

        count = 0
        for node in list_for_each_entry(normal_list, 'struct container',
                                        'list'):
            count += 1

        self.assertTrue(count == expected_count)

    def test_normal_container_list_with_type(self):
        normal_list = get_symbol("good_container_list")
        short_list = get_symbol("good_containers")
        expected_count = short_list.type.sizeof // short_list[0].type.sizeof
        struct_container = gdb.lookup_type('struct container')

        count = 0
        for node in list_for_each_entry(normal_list, struct_container, 'list'):
            count += 1

    def test_cycle_container_list_with_string(self):
        cycle_list = get_symbol("cycle_container_list")
        short_list = get_symbol("cycle_containers")
        expected_count = short_list.type.sizeof // short_list[0].type.sizeof

        count = 0
        with self.assertRaises(ListCycleError):
            for node in list_for_each_entry(cycle_list, 'struct container',
                                            'list', exact_cycles=True,
                                            print_broken_links=False):
                count += 1

    def test_cycle_container_list_with_type(self):
        cycle_list = get_symbol("cycle_container_list")
        short_list = get_symbol("cycle_containers")
        expected_count = short_list.type.sizeof // short_list[0].type.sizeof
        struct_container = gdb.lookup_type('struct container')

        count = 0
        with self.assertRaises(ListCycleError):
            for node in list_for_each_entry(cycle_list, struct_container,
                                            'list', exact_cycles=True,
                                            print_broken_links=False):
                count += 1

    def test_bad_container_list_with_string(self):
        bad_list = get_symbol("bad_container_list")
        short_list = get_symbol("bad_containers")
        expected_count = short_list.type.sizeof // short_list[0].type.sizeof

        count = 0
        with self.assertRaises(CorruptListError):
            for node in list_for_each_entry(bad_list, 'struct container',
                                            'list', print_broken_links=False):
                count += 1

    def test_bad_container_list_with_type(self):
        bad_list = get_symbol("bad_container_list")
        short_list = get_symbol("bad_containers")
        expected_count = short_list.type.sizeof // short_list[0].type.sizeof
        struct_container = gdb.lookup_type('struct container')

        count = 0
        with self.assertRaises(CorruptListError):
            for node in list_for_each_entry(bad_list, struct_container,
                                            'list', print_broken_links=False):
                count += 1
